### DISCUSSION TITLE
how to solve this “No connection to pgmoneta on /tmp”

### DISCUSSION THREAD

**Discussion by skpupil** [pgmoneta/pgmoneta#59] [General]
the pg_hba.conf is 
 ```
local   all             all                                     peer
host    all             all             127.0.0.1/32            trust
host    all             all             ::1/128                 trust
local   replication     all                                     peer
host    replication     all             127.0.0.1/32            ident
host    replication     all             ::1/128                 ident
local   postgres        repl                                   scram-sha-256
host    postgres        repl           127.0.0.1/32            scram-sha-256
host    postgres        repl           ::1/128                 scram-sha-256
host    replication     repl           127.0.0.1/32            scram-sha-256
host    replication     repl           ::1/128                 scram-sha-256
```
I run this
```
/usr/pgsql-10/bin/createdb -U postgres -h localhost -p 5432 kklx
```
In postgres I run this 
```
SELECT pg_create_physical_replication_slot('repl');
CREATE ROLE repl WITH LOGIN REPLICATION PASSWORD 'abc123';
```

In postgres.conf I set this `unix_socket_directories = '/tmp'`

the pgmoneta.conf
```
[pgmoneta]
host = *
metrics = 5001

base_dir = /home/kk/GSoC/pgmoneta/

compression = zstd

retention = 7

log_type = file
log_level = info
log_path = /tmp/pgmoneta.log

unix_socket_dir = /tmp
#unix_socket_dir = /var/run/postgresql
pgsql_dir = /usr/bin/

[kklx]
host = localhost
port = 5432
user = repl
wal_slot = repl
```
I find something under ``/tmp``: 
```
[root@localhost kk]# find /tmp  -name '*5432*' 
/tmp/.s.PGSQL.5432.lock
/tmp/.s.PGSQL.5432
```

When I run ``./pgmoneta-cli -c ./configs/pgmoneta.conf backup primary`` , I met this problem ``No connection to pgmoneta on /tmp``

**Comment by skpupil** [comment]
When I run ```./pgmoneta-cli -c ./configs/pgmoneta.conf backup kklx``` ,I still met this problem.

**Comment by jesperpedersen** [comment]
Questions should be posted in the "General" discussion forum.

The Unix Domain Socket is located in your example @ `/tmp/.s.pgmoneta`. Are you seeing any issues during startup ? What is the filesystem for `/tmp` ?

**Comment by skpupil** [comment]
I cannot find the``/tmp/.s.pgmoneta``. Perhaps pgmoneta has not run successfully, and this file has not been created.  Or do you mean i should modify to ``unix_socket_dir = /tmp/.s.pgmoneta``? I tried. It did not work.
The filesystem of ``/tmp`` is ``xfs``.
```
[root@localhost kk]# df -T
Filesystem          Type     1K-blocks    Used Available Use% Mounted on
devtmpfs            devtmpfs   2829420       0   2829420   0% /dev
tmpfs               tmpfs      2858916     100   2858816   1% /dev/shm
tmpfs               tmpfs      2858916   26428   2832488   1% /run
tmpfs               tmpfs      2858916       0   2858916   0% /sys/fs/cgroup
/dev/mapper/rl-root xfs       31961860 6448368  25513492  21% /
/dev/nvme0n1p1      xfs        1038336  321788    716548  31% /boot
tmpfs               tmpfs       571780      36    571744   1% /run/user/1000
```

**Reply by jesperpedersen** [reply]
You should just use `/tmp` and XFS is ok.

If it fails during startup there should be a `FATAL` log for it.

**Comment by skpupil** [comment]
Startup of pgmoneta? Or postgres?
Where the location of log file?

**Reply by jesperpedersen** [reply]
pgmoneta - and you did `/tmp/pgmoneta.log`

**Comment by skpupil** [comment]
``/tmp/pgmoneta.log`` is empty. 
It really confused me.

**Reply by jesperpedersen** [reply]
Check if there is some output on the terminal during startup

**Reply by skpupil** [reply]
There is nothing abnormal when I run 
```
cmake -DCMAKE_BUILD_TYPE=Debug ..
make
```
What's the meaning of startup.

**Reply by jesperpedersen** [reply]
`./pgmoneta -c ... -u ...`

**Reply by skpupil** [reply]
There is no error when i run ``./pgmoneta -c ./configs/pgmoneta.conf ``.
I don't know how to write``pgmoneta_users.conf``. Because there is  no example  in the ``Getting started with pgmoneta``.

**Reply by jesperpedersen** [reply]
Its in the last section before "Running"

**Reply by skpupil** [reply]
I got this ``2022-05-30 20:39:00 ERROR wal.c:134 WAL: Server kklx is not in a valid configuration
``. How can i solve it?

**Reply by jesperpedersen** [reply]
Did you change `wal_level` to replica ?

**Reply by skpupil** [reply]
How do this change?
In postgresql.conf ,
```
# - Settings -

wal_level = replica                     # minimal, replica, or logical
                                        # (change requires restart)
```
then run ``/usr/pgsql-10/bin/pg_ctl restart``
but still
```
2022-05-30 20:55:00 ERROR server.c:117 Server kklx need wal_level at replica or logical
2022-05-30 20:55:00 ERROR wal.c:134 WAL: Server kklx is not in a valid configuration
```

**Reply by jesperpedersen** [reply]
Did you verify access ? `psql -h localhost -p 5432 -U repl postgres`

**Reply by skpupil** [reply]
I set like this, others are commented by me:
```
local   postgres        repl                                   scram-sha-256
host    postgres        repl           127.0.0.1/32            scram-sha-256
host    postgres        repl           ::1/128                 scram-sha-256
host    replication     repl           127.0.0.1/32            scram-sha-256
host    replication     repl           ::1/128                 scram-sha-256
```
and 
```
[postgres@localhost data]$ psql -h localhost -p 5432 -U repl postgres
Password for user repl: 
psql: FATAL:  password authentication failed for user "repl"
[postgres@localhost data]$ psql -U postgres -h localhost -p 5432
psql (10.21)
Type "help" for help.

postgres=# \q
[postgres@localhost data]$
```

**Reply by jesperpedersen** [reply]
The `repl` user needs access to `postgres` for it to work. Wrong password ?

Check your `password_encryption` and align them between PostgreSQL and pgmoneta - maybe recreate the user. Remember to recreate `pgmoneta_users.conf`

**Reply by skpupil** [reply]
I set like this(add the first line): 
```
host    all             all             ::1/128                 password
local   postgres        repl                                   scram-sha-256
host    postgres        repl           127.0.0.1/32            scram-sha-256
host    postgres        repl           ::1/128                 scram-sha-256
host    replication     repl           127.0.0.1/32            scram-sha-256
host    replication     repl           ::1/128                 scram-sha-256
```
I tried again:
```
[postgres@localhost data]$ psql -h localhost -p 5432 -U repl postgres
Password for user repl: 
psql (10.21)
Type "help" for help.

postgres=> \q
[postgres@localhost data]$ 
```
But:
```
[kk@localhost src]$ ./pgmoneta -c ./configs/pgmoneta.conf  -u ./configs/pgmoneta_users.conf 
pg_receivewal: FATAL:  password authentication failed for user "repl"
pg_receivewal: FATAL:  password authentication failed for user "repl"
```

**Reply by skpupil** [reply]
What's your ``pg_hba.conf``? I'm really confused.

**Comment by skpupil** [comment]
I tried ``wal_level = logical                     # minimal, replica, or logical`` and ````wal_level = replica                     # minimal, replica, or logical````. It didnot work:
```
2022-05-31 08:13:00 ERROR server.c:117 Server kklx need wal_level at replica or logical
2022-05-31 08:13:00 ERROR wal.c:134 WAL: Server kklx is not in a valid configuration
```
I think this is a bug of pgmoneta.

**Comment by MariamFahmy98** [comment]
Hi @skpupil 

I was having this problem when I tried to build it for the first time, and It was because of an incorrect step I made while building it.

With that being said, Let me revise with you the required steps to build it correctly:

**In a new terminal:**
1- Login as a superuser: ```sudo su - postgres```
2- Then we can get to postgresql shell by using following command: ```psql```
3- use the ```\du``` to list all user accounts (or roles) in the current PostgreSQL database server, the output must contain a role name called ```repl```, Here is a snapshot of my output:
![Screenshot from 2022-05-31 07-27-02](https://user-images.githubusercontent.com/55502281/171100761-e02de5c9-2862-4d1d-8b6a-dba2656da4d7.png)
In case you haven't found it, you can create it by ```CREATE ROLE repl WITH LOGIN REPLICATION PASSWORD 'secretpassword';``` as mentioned in the GETTING_STARTED.md


**In a new terminal:** _Navigate to your pgmoneta directory and open a terminal there_
Here is a snapshot for my pgmoneta directory:
![Screenshot from 2022-05-31 07-27-53](https://user-images.githubusercontent.com/55502281/171101327-148cbc18-2482-4edf-a8ce-b39e332c91fe.png)
It contains the **forked pgmoneta repository**, **a backup directory** (I manually created it to hold all backups), a **pgmoneta.conf** file ( I created it too) and **pgmoneta_users.conf** (It was automatically created when adding ```repl``` as a new user as mentioned in GETTING_STARTED.md)

Here is the content of my **pgmoneta.conf** file:
```
[pgmoneta]
host = *
metrics = 5001

base_dir = backup-directory

compression = lz4

retention = 7

log_type = file
log_level = info
log_path = /tmp/pgmoneta.log

unix_socket_dir = /tmp/
pgsql_dir = /usr/bin/

[primary]
host = localhost
port = 5432
user = repl
```
If all the above is done correctly, you can run ```pgmoneta -c pgmoneta.conf -u pgmoneta_users.conf``` and in a new terminal you can do a backup by running: ```pgmoneta-cli -c pgmoneta.conf backup primary```

**Note: you must run it in your pgmoneta directory that contains both pgmoneta.conf and pgmoneta_users.conf**

If I were you, I'd go through each step again carefully.

I hope it helps, feel free to ask me and I will help as much as I can.

**Reply by skpupil** [reply]
I have done those you said correctly.
![image](https://user-images.githubusercontent.com/56211074/171175721-0c27a72f-4893-42ca-bfc4-3a3631afb30f.png)
Then I run:
``./pgmoneta-admin master-key`` ask me input something. I input:``postgres``
``./pgmoneta-admin -f ./configs/pgmoneta_users.conf add-user``. I input ``repl`` and verify password twice.
I created ``./configs/pgmoneta.conf``:
```
[pgmoneta]
host = *
metrics = 5001

base_dir = /home/kk/GSoC/pgmoneta/

compression = zstd

retention = 7

log_type = file
log_level = info
log_path = /tmp/pgmoneta.log

unix_socket_dir = /tmp/
#unix_socket_dir = /var/run/postgresql
pgsql_dir = /usr/pgsql-10/bin

[kklx]
host = localhost
port = 5432
user = repl
```
``./pgmoneta -c ./configs/pgmoneta.conf  -u ./configs/pgmoneta_users.conf ``, and I see /tmp/pgmoneta.log:
```
2022-05-31 20:35:35 INFO  main.c:609 pgmoneta: started on *
2022-05-31 20:35:35 ERROR server.c:117 Server kklx need wal_level at replica or logical
2022-05-31 20:35:35 ERROR wal.c:134 WAL: Server kklx is not in a valid configuration
2022-05-31 20:36:00 ERROR server.c:117 Server kklx need wal_level at replica or logical
2022-05-31 20:36:00 ERROR wal.c:134 WAL: Server kklx is not in a valid configuration
```

**Reply by skpupil** [reply]
And the version of postgres is 10.21.

**Comment by jesperpedersen** [comment]
I have added https://github.com/pgmoneta/pgmoneta/blob/main/doc/tutorial/01_install.md

If you run it in a docker/podman environment look at https://github.com/jesperpedersen/podman-pgsql/tree/master/pgsql14-pgmoneta-rocky8 in addition

**Reply by skpupil** [reply]
It's so kind of you. I tied again with your document. It worked, and the wired problem dissapeared!

**Comment by jesperpedersen** [comment]
Good that is working for you now - so I have closed the tutorial issues for `01_install.md` and `02_replication_slot.md`.
